domain/port/state/reconcile.go

```diff
@@ -0,0 +1,54 @@
+// Copyright 2024 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package state
+
+import (
+	"github.com/juju/juju/core/interval"
+	"github.com/juju/juju/core/network"
+)
+
+func reconcilePorts(currentOpened, openPorts, closePorts []network.PortRange) []network.PortRange {
+	reconciled := map[string]interval.IntegerIntervals{
+		"tcp": interval.NewIntegerIntervals(),
+		"udp": interval.NewIntegerIntervals(),
+	}
+	var icmp bool
+
+	for _, port := range append(currentOpened, openPorts...) {
+		if port.Protocol == "icmp" {
+			icmp = true
+			continue
+		}
+		reconciled[port.Protocol] = reconciled[port.Protocol].Union(interval.NewIntegerInterval(port.FromPort, port.ToPort))
+	}
+
+	for _, port := range closePorts {
+		if port.Protocol == "icmp" {
```

Comment:
Could do with a comment to explain this special case.

Reply:
Added